File: experiment-15-sonnet5/brazil/runs/language=typescript_model=opus-4.8_prompt=bdd/rep1/tests/stats.test.ts
import { describe, it, expect } from "vitest";
import { aggregateStats, biggestWins, topScoringTeams } from "../src/queries/stats.js";
import { makeMatch, realStore } from "./fixtures.js";

describe("Aggregate match statistics", () => {
  it("given three matches, when aggregated, then average goals per match is correct", () => {
    // Given: total 6 goals over 3 matches
    const matches = [
      makeMatch({ homeGoals: 2, awayGoals: 1 }),
      makeMatch({ homeGoals: 1, awayGoals: 1 }),
      makeMatch({ homeGoals: 0, awayGoals: 1 }),
    ];
    // When
    const agg = aggregateStats(matches);
    // Then
    expect(agg.averageGoalsPerMatch).toBe(2);
  });

  it("given a home win and an away win, when aggregated, then win rates split evenly", () => {
    // Given
    const matches = [
      makeMatch({ homeGoals: 3, awayGoals: 0 }),
      makeMatch({ homeGoals: 0, awayGoals: 2 }),
    ];
    // When
    const agg = aggregateStats(matches);
    // Then
    expect(agg.homeWins).toBe(1);
    expect(agg.awayWins).toBe(1);
    expect(agg.homeWinRate).toBeCloseTo(0.5);
  });
});

describe("Biggest wins", () => {
  it("given matches of varied margins, when ranked, then the largest margin comes first", () => {
    // Given
    const matches = [
      makeMatch({ homeTeam: "A", awayTeam: "B", homeGoals: 2, awayGoals: 1 }),
      makeMatch({ homeTeam: "C", awayTeam: "D", homeGoals: 8, awayGoals: 0 }),
    ];
    // When
    const wins = biggestWins(matches, {}, 5);
    // Then
    expect(wins[0].margin).toBe(8);
    expect(wins[0].match.homeTeam).toBe("C");
  });
});

describe("Top scoring teams", () => {
  it("given matches, when totalled, then the highest-scoring team ranks first", () => {
    // Given: A scores 3+2, B scores 0+1
    const matches = [
      makeMatch({ homeTeam: "A", awayTeam: "B", homeGoals: 3, awayGoals: 0 }),
      makeMatch({ homeTeam: "B", awayTeam: "A", homeGoals: 1, awayGoals: 2 }),
    ];
    // When
    const totals = topScoringTeams(matches, {}, 5);
    // Then
    expect(totals[0].team).toBe("A");
    expect(totals[0].goalsFor).toBe(5);
  });
});

describe("Aggregate statistics over real data", () => {
  it("given the real Brasileirão, when aggregated, then average goals per match is football-plausible", () => {
    // Given the loaded datasets
    const store = realStore();
    // When
    const agg = aggregateStats(store.matches, { competition: "Brasileirão" });
    // Then: real leagues sit roughly between 2 and 3 goals per game
    expect(agg.averageGoalsPerMatch).toBeGreaterThan(2);
    expect(agg.averageGoalsPerMatch).toBeLessThan(3.5);
  });
});

File: experiments/adrianco/experiment-46-opus5/brazil/runs/language=elixir_model=claude-opus-5_prompt=neutral/rep1/test/query/stats_test.exs
defmodule BrazilianSoccer.Query.StatsTest do
  @moduledoc """
  Feature: Statistical analysis

  Scenario: Aggregate statistics
    Given the match data is loaded
    When I ask for goals per match, home advantage or biggest wins
    Then the numbers are computed over every matching match, not a page of them
  """

  use BrazilianSoccer.GraphCase, async: true

  describe "Scenario: overview statistics" do
    test "Given a competition Then goals and result rates are consistent", %{graph: graph} do
      result = ok!(Stats.overview(graph, %{competition: "Brasileirão"}))

      assert result.matches > 8_000
      assert result.matches == result.home_wins + result.away_wins + result.draws
      assert_in_delta result.home_win_rate + result.away_win_rate + result.draw_rate, 100.0, 0.2
      assert_in_delta result.goals_per_match, 2.5, 0.5
      assert result.goals == result.home_goals + result.away_goals
    end

    test "Given the whole graph Then every competition shows up in the breakdown", %{graph: graph} do
      result = ok!(Stats.overview(graph, %{}))
      competitions = Enum.map(result.by_competition, fn {competition, _} -> competition end)

      assert :serie_a in competitions
      assert :libertadores in competitions
      assert result.matches > 15_000
    end

    test "Given a season range Then only those seasons are aggregated", %{graph: graph} do
      result =
        ok!(
          Stats.overview(graph, %{competition: "Brasileirão", season_from: 2018, season_to: 2019})
        )

      seasons = Enum.map(result.by_season, fn {season, _} -> season end)
      assert seasons == [2018, 2019]
      assert result.matches == 760
    end

    test "Given a team Then only its matches count", %{graph: graph} do
      {:ok, team} = Graph.find_team(graph, "Flamengo")
      result = ok!(Stats.overview(graph, %{team: "Flamengo"}))

      assert result.matches <= team.match_count
    end

    test "Given filters that match nothing Then an error explains", %{graph: graph} do
      assert {:error, {:no_matches, _}} =
               Stats.overview(graph, %{competition: "Libertadores", season: 1999})
    end
  end

  describe "Scenario: biggest wins" do
    test "Given no filters Then the biggest margins in the whole graph come back", %{graph: graph} do
      result = ok!(Stats.biggest_wins(graph, %{limit: 10}))

      margins = Enum.map(result.matches, &Match.margin/1)
      assert margins == Enum.sort(margins, :desc)
      assert hd(margins) >= 8
      assert length(result.matches) == 10
    end

    test "Given a team Then only wins by that team come back", %{graph: graph} do
      result = ok!(Stats.biggest_wins(graph, %{team: "Palmeiras", limit: 5}))

      assert Enum.all?(result.matches, &(Match.outcome_for(&1, result.team.id) == :win))
    end

    test "Given a competition and season Then the search is scoped", %{graph: graph} do
      result =
        ok!(Stats.biggest_wins(graph, %{competition: "Brasileirão", season: 2019, limit: 3}))

      assert Enum.all?(result.matches, &(&1.competition == :serie_a and &1.season == 2019))
    end
  end

  describe "Scenario: highest scoring matches" do
    test "Given a competition Then matches are ranked by total goals", %{graph: graph} do
      result = ok!(Stats.highest_scoring(graph, %{competition: "Brasileirão", limit: 5}))
      totals = Enum.map(result.matches, &Match.total_goals/1)

      assert totals == Enum.sort(totals, :desc)
      assert hd(totals) >= 8
    end
  end

  describe "Scenario: comparing seasons" do
    test "Given two seasons Then each gets its own block with a champion", %{graph: graph} do
      result =
        ok!(Stats.compare_seasons(graph, %{competition: "Brasileirão", seasons: [2018, 2019]}))

      assert Enum.map(result.seasons, & &1.season) == [2018, 2019]
      assert Enum.all?(result.seasons, & &1.available)
      assert Enum.map(result.seasons, & &1.champion.name) == ["Palmeiras", "Flamengo"]
      assert Enum.all?(result.seasons, &(&1.stats.matches == 380))
      assert Enum.all?(result.seasons, &(&1.top_scoring_team.goals > 50))
    end

    test "Given a season with no data Then it is reported as unavailable", %{graph: graph} do
      result = ok!(Stats.compare_seasons(graph, %{seasons: [2019, 1990]}))

      assert Enum.find(result.seasons, &(&1.season == 1990)).available == false
    end

    test "Given no seasons Then the error says they are required", %{graph: graph} do
      assert {:error, {:missing_argument, :seasons}} = Stats.compare_seasons(graph, %{})
    end
  end

  describe "Scenario: home advantage" do
    test "Given a competition Then home advantage is reported per season", %{graph: graph} do
      result = ok!(Stats.home_advantage(graph, %{competition: "Brasileirão"}))

      assert result.overall.home_win_rate > result.overall.away_win_rate
      assert length(result.by_season) == 21

      assert Enum.all?(result.by_season, fn {_season, stats} ->
               stats.home_win_rate >= 0 and stats.home_win_rate <= 100
             end)
    end

    test "Given the Libertadores Then home advantage is also computable", %{graph: graph} do
      result = ok!(Stats.home_advantage(graph, %{competition: "Libertadores"}))

      assert result.overall.matches > 1_000
      assert result.overall.home_goals > result.overall.away_goals
    end
  end
end
